### DISCUSSION TITLE
A few thoughts from a PCB maker

### DISCUSSION THREAD

**Discussion by 74HC14** [sn4k3/UVtools#592] [General]
I'll be making solder mask templates soon, and I don't think the maximum 200s is enough to cure my mask in one go. Partly because my LED driver broke and has been replaced with a homebrew circuit with anaemic output (for its own protection). This may apply to people making silkscreens also. 200s is the same time limit I remember seeing on the Anycubic software, I wonder if it's some sort of limitation?

I'd like it if it were possible to just stop the stepper motor from doing anything at all when making PCBs, but maybe that's in the domain of custom printer firmware instead.

Also be nice if gerber outputs from KiCAD could have holes on the copper layers. Not just for drilling after exposing, but also for exposing after drilling. I'll have to look into how KiCAD generates its gerbers to see if it's possible to create holes on them (why are "drill marks" disabled only for gerbers?). Or else use a script to modify the board layout before generating gerbers to remove copper circles mirroring holes. To allow alignment without exposing my resist, I'll see if I can add some red/yellow/green lights between the UV LEDs so I can see the edge-cuts layer and/or holes. Not sure if there's any room for activating that automatically without custom firmware, so I think I'll just add an SPDT switch for it on the side of the printer.

Nice software!

**Comment by sn4k3** [comment]
> I don't think the maximum 200s is enough to cure my mask in one go

You can duplicate the layer the times you want, so you will have 200s x your clones
BUT you can set exposure time to very high value if you use UVtools File -> Terminal and write: `SlicerFile.BottomExposureTime = SlicerFile.ExposureTime = 1000;`

> I wonder if it's some sort of limitation?

It must have have a limit, no one can tell since firmwares are closed and brand will not anwser you correctly even if you ask

> I'd like it if it were possible to just stop the stepper motor from doing anything at all when making PCBs, but maybe that's in the domain of custom printer firmware instead.

Most printers will not allow you to control such, you would need an gcoded printer, eg: IBEE, GKone, NovaMaker

> Also be nice if gerber outputs from KiCAD could have holes on the copper layers.

You asking to subtract holes from copper on UVtools?
Drill file use another language/syntax than gerber

> I'll see if I can add some red/yellow/green lights between the UV LEDs so I can see the edge-cuts layer and/or holes. Not sure if there's any room for activating that automatically without custom firmware, so I think I'll just add an SPDT switch for it on the side of the printer.

Just detect the UV LED trigger by pass it outout to a opto-isolator and use an MCU to trigger your custom LED when UV LED is off

**Reply by 74HC14** [reply]
Just noticed the exposure limit alteration in the changelog, thanks a bunch!

> Drill file use another language/syntax than gerber

Yes but in KiCAD you can output SVGs and PDFs and such, in the exact same window as the gerber export, but with the option to add full-sized or small drill marks. Presumably for the purpose of home etching via toner transfer or transparency, considering those small marks are often used to help drilling. It may be worth me taking the time to output as a .svg and import that into UVtools, not that I know how. I've seen other people with the same problem combine the drill and gerber file, then turn the result back into a gerber. I should be able to make a python script that does that reasonably easily.

> you would need an gcoded printer, eg: IBEE, GKone, NovaMaker

Yeah I figure it wouldn't be easy. I'll see about bodging an MCU in there to detect stepper motor signals, and instantly trigger the depth sensor.

> Just detect the UV LED trigger by pass it outout to a opto-isolator and use an MCU to trigger your custom LED when UV LED is off

Would the LCD be showing anything other than a blank screen when the UV LEDs are off? Hard to see I guess. I'll get some RGB LED strips tomorrow and check it, though the lenses may make getting light to shine straight up kinda difficult. Perhaps I'll have to make a custom PCB to hold SMD LEDs as close to the existing UV LEDs as possible...

**Reply by sn4k3** [reply]
About the drill stuff, just make two pictures, one without the drill marks and other with drill marks using the UVtools generated image so i can understand what you are trying to achieve.

> Would the LCD be showing anything other than a blank screen when the UV LEDs are off? 

Depends on printers, some just leave last image in because blanking screen takes time, so it not worth to clear image, it just replace old image with new image when required and it spare a draw.
Few will blank image after LED off.

**Reply by 74HC14** [reply]
> About the drill stuff, just make two pictures, one without the drill marks and other with drill marks using the UVtools generated image so i can understand what you are trying to achieve.

Can't figure out how to get the SVG into UVtools, but here's what I could do:

<a href="https://anonfiles.com/S7GcodF8y6/SS_2022-10-31_at_10.29.18_PM_png">A screenshot of the SVG file with its small holes for drilling.</a>
![image](https://user-images.githubusercontent.com/113281/199029296-94e975a2-8263-40fd-83a9-e0ed5a2051a6.png)



<a href="https://anonfiles.com/VeGdo5Fby1/SS_2022-10-31_at_10.30.32_PM_png">A screenshot of KiCAD's SVG export window, next to the board layout showing full-sized holes.</a>
![image](https://user-images.githubusercontent.com/113281/199029359-f60e6ae1-de3e-4488-9496-3eda437cf020.png)

<a href="https://anonfiles.com/U0G6odFayc/SS_2022-10-31_at_10.37.23_PM_png">The copper layer gerber imported into UVtools, with lack of holes.</a>
![image](https://user-images.githubusercontent.com/113281/199029423-61353430-f0b7-4cde-90c9-313e05c11195.png)

<a href="https://anonfiles.com/T3Gbo1F6y5/SS_2022-10-31_at_10.37.46_PM_png">The KiCAD Gerber export window, showing the "Drill marks" option greyed out and unusable for whatever reason.</a>
![image](https://user-images.githubusercontent.com/113281/199029595-2481ae86-cba3-4a16-84fe-2299addefabe.png)

Looking at the files, both KiCAD's X2 and Excellon drill files seem to be reasonably readable, so I should be able to write a script to grab the hole coordinates (and diameters if I want full-size holes) and add them to a copper gerber file. The X2 ones are in gerber format so they can be imported into UVtools, but from what I see there's no way of subtracting gerbers instead of adding them.

> some just leave last image in because blanking screen takes time

Oh I guess that works.

**Reply by sn4k3** [reply]
> Can't figure out how to get the SVG into UVtools, but here's what I could do:

You can't. You can only import image file types (PNG). But it's possible via scripting.
I can add that latter too.

> The X2 ones are in gerber format so they can be imported into UVtools, but from what I see there's no way of subtracting gerbers instead of adding them.

1. Import both gerbers by selecting the two files (will create two layers)
2. Go to tools - Layer arithmetic
3. (If hole layer is index 1) Write: `0-1`
3. (If hole layer is index 0) Write: `1=1-0`
4. Remove hole layer (CTRL + DELETE)
5. Done

But now i see what you trying.
Send me a zip file with the gerber(s) and hole file(s) so i can test some methods

**Reply by sn4k3** [reply]
I've been playing with the PCB tool, see if that changes fits your needs:

Copper:
![image](https://user-images.githubusercontent.com/113281/199093250-e567021e-b946-44b6-9e38-a992bd9d3951.png)

Drills as gerber:
![image](https://user-images.githubusercontent.com/113281/199093293-b6ae115f-0ee9-4721-90d1-4d4b153ae5a3.png)

Result:
![image](https://user-images.githubusercontent.com/113281/199093323-5bc055c5-2631-4d08-8569-3fb74e64f88d.png)

In the end it produce same effect as the steps I wrote above with "Layer Arithmetic"

**Reply by 74HC14** [reply]
Ah yeah that layer arithmetic works great. Being able to invert files and add them as they're imported like in your picture would be even easier. You're doing great work here, thanks a bunch!
<img width="511" alt="SS 2022-11-01 at 8 38 24 PM" src="https://user-images.githubusercontent.com/40238857/199182917-50600e85-d156-4df7-a027-98fe637bde21.png">

For some reason github is preventing me from uploading, so here's the gerbers on anonfiles:
https://anonfiles.com/M2I6veF3yc/gerbers_zip

**Reply by sn4k3** [reply]
> https://anonfiles.com/M2I6veF3yc/gerbers_zip

Ok, that works in my implementation 👍
Will be on next release.

**Comment by stonemull** [comment]
I just want to thank you for doing this. Only discovered uvtools could do this yesterday as it does not appear to be common knowledge yet. I will be telling people. It’s awesome !
